// @umituz/react-native-design-system/src/theme/core/colors/DarkColors.ts
/**
 * DARK THEME COLORS
 *
 * Dark theme color palette — Forest Green & Warm Orange (MoveLog theme)
 */

export const darkColors = {
  // PRIMARY BRAND COLORS - Forest Green (lighter for dark backgrounds)
  primary: '#5AAF7F',          // Lighter Forest Green for dark mode
  primaryLight: '#6BC48F',     // Light Green
  primaryDark: '#4A9B6F',      // Forest Green

  secondary: '#FF8C42',        // Warm Orange
  secondaryLight: '#FFA06A',   // Light Orange
  secondaryDark: '#E57030',    // Dark Orange

  accent: '#FF8C42',           // Warm Orange
  accentLight: '#FFA06A',      // Light Orange
  accentDark: '#E57030',       // Dark Orange

  // MATERIAL DESIGN 3 - ON COLORS (Dark mode)
  onPrimary: '#000000',
  onSecondary: '#000000',
  onSuccess: '#000000',
  onError: '#FFFFFF',
  onWarning: '#000000',
  onInfo: '#000000',
  onSurface: '#E2E8F0',
  onBackground: '#F1F5F9',
  onSurfaceDisabled: '#6C757D',
  onSurfaceVariant: '#98989D',

  // MATERIAL DESIGN 3 - CONTAINER COLORS (Dark mode)
  primaryContainer: '#2D6B4A',    // Dark green container
  onPrimaryContainer: '#D1F5E3',  // Light green text
  secondaryContainer: '#CC5500',  // Dark orange container
  onSecondaryContainer: '#FFE8D4', // Light text
  errorContainer: '#7F1D1D',
  onErrorContainer: '#FEE2E2',

  // MATERIAL DESIGN 3 - OUTLINE (Dark mode)
  outline: '#48484A',
  outlineVariant: '#3A3A3C',
  outlineDisabled: '#3A3A3C',

  // SEMANTIC UI COLORS (slightly lighter for dark backgrounds)
  success: '#5AAF7F',             // Lighter green for dark mode
  successLight: '#6BC48F',
  successDark: '#4A9B6F',

  error: '#EF4444',
  errorLight: '#F87171',
  errorDark: '#DC2626',

  warning: '#FFC107',
  warningLight: '#FFD54F',
  warningDark: '#E6A800',

  info: '#FF8C42',
  infoLight: '#FFA06A',
  infoDark: '#E57030',

  // SEMANTIC CONTAINER COLORS
  successContainer: '#D1F5E3',
  onSuccessContainer: '#2D6B4A',
  warningContainer: '#FEF3C7',
  onWarningContainer: '#92400E',
  infoContainer: '#FFE8D4',
  onInfoContainer: '#CC5500',

  // GRAYSCALE PALETTE
  gray50: '#FAFAFA',
  gray100: '#F4F4F5',
  gray200: '#E4E4E7',
  gray300: '#D4D4D8',
  gray400: '#A1A1AA',
  gray500: '#71717A',
  gray600: '#52525B',
  gray700: '#3F3F46',
  gray800: '#27272A',
  gray900: '#18181B',

  // BACKGROUND COLORS — MoveLog's signature dark palette
  backgroundPrimary: '#1C1C1E',      // MoveLog dark root
  backgroundSecondary: '#2C2C2E',    // MoveLog dark default

  surface: '#2C2C2E',                // MoveLog backgroundDefault
  surfaceVariant: '#3A3A3C',         // MoveLog backgroundSecondary
  surfaceSecondary: '#3A3A3C',
  surfaceDisabled: '#48484A',        // MoveLog backgroundTertiary

  // TEXT COLORS (dark mode)
  textPrimary: '#FFFFFF',            // MoveLog dark text
  textSecondary: '#98989D',          // MoveLog dark textSecondary
  textTertiary: '#6C757D',
  textDisabled: '#48484A',
  textInverse: '#1C1C1E',

  // BORDER COLORS (dark mode)
  border: '#48484A',                 // MoveLog dark border
  borderLight: '#3A3A3C',
  borderMedium: '#636366',
  borderFocus: '#5AAF7F',            // Green focus
  borderDisabled: '#3A3A3C',

  // COMPONENT-SPECIFIC COLORS (dark mode)
  buttonPrimary: '#5AAF7F',
  buttonSecondary: '#FF8C42',

  inputBackground: '#2C2C2E',
  inputBorder: '#48484A',

  cardBackground: '#2C2C2E',

  // COLOR ALIASES
  text: '#FFFFFF',
  background: '#1C1C1E',
  card: '#2C2C2E',

  // SPECIAL COLORS
  transparent: 'transparent',
  black: '#000000',
  white: '#FFFFFF',

  // RGBA OVERLAY COLORS
  modalOverlay: 'rgba(0, 0, 0, 0.7)',
  overlaySubtle: 'rgba(0, 0, 0, 0.05)',
  overlayLight: 'rgba(0, 0, 0, 0.1)',
  overlayMedium: 'rgba(0, 0, 0, 0.4)',
  overlayBackground: 'rgba(0, 0, 0, 0.1)',

  whiteOverlay: 'rgba(255, 255, 255, 0.2)',
  whiteOverlayStrong: 'rgba(255, 255, 255, 0.95)',
  whiteOverlayBorder: 'rgba(255, 255, 255, 0.3)',

  textWhiteOpacity: 'rgba(255, 255, 255, 0.8)',

  errorBackground: 'rgba(239, 68, 68, 0.1)',
  primaryBackground: 'rgba(90, 175, 127, 0.15)',  // Green overlay (dark mode)

  cardOverlay: 'rgba(0, 0, 0, 0.25)',

  inputBackground_RGBA: 'rgba(44, 44, 46, 0.9)',
};
