// @umituz/react-native-design-system/src/theme/core/colors/LightColors.ts
/**
 * LIGHT THEME COLORS
 *
 * Light theme color palette — Forest Green & Warm Orange (MoveLog theme)
 */

export const lightColors = {
  // PRIMARY BRAND COLORS - Forest Green (MoveLog)
  primary: '#4A9B6F',          // Forest Green
  primaryLight: '#6BC48F',     // Light Green
  primaryDark: '#3A7A5A',      // Deep Green

  secondary: '#FF8C42',        // Warm Orange (accent)
  secondaryLight: '#FFA06A',   // Light Orange
  secondaryDark: '#E57030',    // Dark Orange

  accent: '#FF8C42',           // Warm Orange
  accentLight: '#FFA06A',      // Light Orange
  accentDark: '#E57030',       // Dark Orange

  // MATERIAL DESIGN 3 - ON COLORS
  onPrimary: '#FFFFFF',
  onSecondary: '#FFFFFF',
  onSuccess: '#FFFFFF',
  onError: '#FFFFFF',
  onWarning: '#000000',
  onInfo: '#FFFFFF',
  onSurface: '#1C1C1E',
  onBackground: '#1C1C1E',
  onSurfaceDisabled: '#C7C7CC',
  onSurfaceVariant: '#6C757D',

  // MATERIAL DESIGN 3 - CONTAINER COLORS
  primaryContainer: '#D1F5E3',    // Light green container
  onPrimaryContainer: '#2D6B4A',  // Dark green text
  secondaryContainer: '#FFE8D4',  // Light orange container
  onSecondaryContainer: '#CC5500', // Dark orange text
  errorContainer: '#FEE2E2',
  onErrorContainer: '#991B1B',

  // MATERIAL DESIGN 3 - OUTLINE
  outline: '#C7C7CC',
  outlineVariant: '#E5E5EA',
  outlineDisabled: '#E5E5EA',

  // SEMANTIC UI COLORS
  success: '#4A9B6F',             // Forest Green (same as primary)
  successLight: '#6BC48F',
  successDark: '#3A7A5A',

  error: '#DC3545',
  errorLight: '#E87070',
  errorDark: '#B02A37',

  warning: '#FFC107',
  warningLight: '#FFD54F',
  warningDark: '#E6A800',

  info: '#FF8C42',                // Warm Orange for info
  infoLight: '#FFA06A',
  infoDark: '#E57030',

  // SEMANTIC CONTAINER COLORS
  successContainer: '#D1F5E3',
  onSuccessContainer: '#2D6B4A',
  warningContainer: '#FFF8DC',
  onWarningContainer: '#92400E',
  infoContainer: '#FFE8D4',
  onInfoContainer: '#CC5500',

  // GRAYSCALE PALETTE
  gray50: '#FAFAFA',
  gray100: '#F4F4F5',
  gray200: '#E5E5EA',
  gray300: '#D1D1D6',
  gray400: '#AEAEB2',
  gray500: '#8E8E93',
  gray600: '#636366',
  gray700: '#48484A',
  gray800: '#3A3A3C',
  gray900: '#1C1C1E',

  // BACKGROUND COLORS — clean iOS-style whites
  backgroundPrimary: '#FFFFFF',
  backgroundSecondary: '#F8F9FA',

  surface: '#FFFFFF',
  surfaceVariant: '#F2F2F7',
  surfaceSecondary: '#F2F2F7',
  surfaceDisabled: '#F4F4F4',

  // TEXT COLORS
  textPrimary: '#1C1C1E',
  textSecondary: '#6C757D',
  textTertiary: '#98989D',
  textDisabled: '#C7C7CC',
  textInverse: '#FFFFFF',

  // BORDER COLORS
  border: '#E5E5EA',
  borderLight: '#F2F2F7',
  borderMedium: '#C7C7CC',
  borderFocus: '#4A9B6F',
  borderDisabled: '#F2F2F7',

  // COMPONENT-SPECIFIC COLORS
  buttonPrimary: '#4A9B6F',
  buttonSecondary: '#FF8C42',

  inputBackground: '#FFFFFF',
  inputBorder: '#E5E5EA',

  cardBackground: '#FFFFFF',

  // COLOR ALIASES
  text: '#1C1C1E',
  background: '#FFFFFF',
  card: '#FFFFFF',

  // SPECIAL COLORS
  transparent: 'transparent',
  black: '#000000',
  white: '#FFFFFF',

  // RGBA OVERLAY COLORS
  modalOverlay: 'rgba(0, 0, 0, 0.5)',
  overlaySubtle: 'rgba(0, 0, 0, 0.05)',
  overlayLight: 'rgba(0, 0, 0, 0.1)',
  overlayMedium: 'rgba(0, 0, 0, 0.3)',
  overlayBackground: 'rgba(0, 0, 0, 0.05)',

  whiteOverlay: 'rgba(255, 255, 255, 0.2)',
  whiteOverlayStrong: 'rgba(255, 255, 255, 0.95)',
  whiteOverlayBorder: 'rgba(255, 255, 255, 0.5)',

  textWhiteOpacity: 'rgba(255, 255, 255, 0.8)',

  errorBackground: 'rgba(220, 53, 69, 0.1)',
  primaryBackground: 'rgba(74, 155, 111, 0.1)',  // Green background

  cardOverlay: 'rgba(0, 0, 0, 0.08)',

  inputBackground_RGBA: 'rgba(248, 249, 250, 0.9)',
};
